src/lib/modes/aead/ccm/ccm.cpp

```diff
@@ -91,18 +91,18 @@ void CCM_Mode::key_schedule(const uint8_t key[], size_t length)
    m_cipher->set_key(key, length);
    }
 
-void CCM_Mode::set_associated_data(const uint8_t ad[], size_t length)
+void CCM_Mode::set_ad_n(size_t /* ignored */, std::span<const uint8_t> ad)
```

Comment:
Ignoring the specified AD index seems like a really bad idea to me. There is a reason the current code throws if you attempt to use an idx > 0 with a mode that doesn't support more than one AD, to do otherwise is to invite really subtle bugs.

Reply:
I introduced `set_ad_n` as the internal interface to get around the virtual dispatch issue. The idea was to combine both virtual methods `set_additional_data` and `set_additional_data_n` for the internal implementation, "saving" one virtual method. The public interface stays the same and throws when misused as before.

Reply:
Done.